yacl/crypto/primitives/tpre/README.md

```diff
@@ -0,0 +1,102 @@
+
+>TPRE is a substitute of [umbral](https://github.com/nucypher/umbral-doc/blob/master/umbral-doc.pdf). It is implemented using C++, and in addition, it replaces algorithms such as ECC and AES with SM2, SM3 and SM4
+
+# 2. Background knowledge
+
+## 2.1 Proxy Re-Encryption
+
+Proxy Re-Encryption is a type of public key cryptography that allows the proxy to convert one public key related ciphertext to another, while the proxy cannot understand any information about the original message; To achieve this, the agent must have a re-encryption key. A Proxy Re-Encryption algorithm typically consists of three roles: data owner Alice, data receiver Bob, and proxy computing proxy. Assuming that the data $m$ has been encrypted into ciphertext $c$ by Alice using her own public key and stored in Proxy, the specific steps of the algorithm are as follows:
+
+1. As the data owner, Alice wants to authorize Bob to use the data $m$, and generates a Re-Encryption key $rk$ for the proxy.
+
+2. After receiving $rk$, the proxy re encrypts the ciphertext $c$ to obtain a new ciphertext $c'$, and send $c'$ to Bob.
+
+3. Bob decrypts $c'$ using his own private key to obtain plaintext data $m$.
+
+## 2.2 Distributed threshold Proxy Re-Encryption
+
+Proxy Re-Encryption is suitable for use in cloud computing scenarios, where the proxy node is a single node with strong computing performance. This is not in line with the existing privacy computing architecture, as it is usually a distributed architecture. Therefore, it is necessary to modify the traditional Proxy Re-Encryption scheme to adapt to distributed computing environments.
+
+Distributed Proxy Re-Encryption refers to splitting a single proxy node in traditional Proxy Re-Encryption into multiple proxy nodes. Therefore, when re encrypting data, multiple proxy nodes are required to participate in collaborative computing.
+
+Considering the flexibility of selecting proxy nodes to participate in calculations, it is necessary to redesign distributed Proxy Re-Encryption to a threshold based distributed Proxy Re-Encryption.
+
+##2.3 Elliptic Curve
+
+The following is the parameter selection for the prime field elliptic curve "sm2p256v1" selected by SM2:
+
+>p = FFFFFFFE FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF 00000000 FFFFFFFF FFFFFFFF
+
+>a = FFFFFFFE FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF 00000000 FFFFFFFF FFFFFFFC
+
+>b = 28E9FA9E 9D9F5E34 4D5A9E4B CF6509A7 F39789F5 15AB8F92 DDBCBD41 4D940E93
+
+>n = FFFFFFFE FFFFFFFF FFFFFFFF FFFFFFFF 7203DF6B 21C6052B 53BBF409 39D54123
+
+>g_ x = 32C4AE2C 1F198119 5F990446 6A39C994 8FE30BBF F2660BE1 715A4589 334C74C7\
+
+>g_ y = BC3736A2 F4F6779C 59BDCEE3 6B692153 D0A9877C C62A4740 02DF32E5 2139F0A0
+
+## 2.3  hash function
+
+Using the hash algorithm, construct the following, where $n$ is the order of the elliptic curve and $x$ is the input to the function:
+
+$$h_ x = 1 + \text{Bignum(SM3}(x)||\text{SM3(SM3(}x))) \bmod n-1$$
```

Comment:
Is there any reason to use the modulus `mod n-1`? `mod n`'s result is just in `[0,n-1]`. Note that the order `n` of elliptic curve(SM2) base on prime filed is also a prime number and so `n-1` is not uncertainly a prime number.

Reply:
This is written as n-1 because it respects Chapter 5.2 of the original paper：https://github.com/nucypher/umbral-doc/blob/master/umbral-doc.pdf，